staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go

```diff
@@ -237,3 +241,81 @@ func newExtra(h http.Header, headerPrefixes []string) map[string][]string {
 
 	return ret
 }
+
+func deserializeStrings(in string) ([]string, error) {
+	if len(in) == 0 {
+		return nil, nil
+	}
+	var ret []string
+	if err := json.Unmarshal([]byte(in), &ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+// TODO: doc
+type DynamicRequestHeaderAuthRequestProvider struct {
+	configmapLister corev1listers.ConfigMapNamespaceLister
+	configmapName   string
+	configmapNamespace string
+	authenticationRoleName string
+
+	usernameHeadersKey     string
+	groupHeadersKey        string
+	extraHeaderPrefixesKey string
+	allowedClientNamesKey  string
+}
+
+// TODO: doc
+func NewDynamicRequestHeaderAuthRequestProvider(configmapLister corev1listers.ConfigMapNamespaceLister, configmapName, configmapNamespace, authenticationRoleName,
+	usernameHeadersKey, groupHeadersKey, extraHeaderPrefixesKey, allowedClientNamesKey string) *DynamicRequestHeaderAuthRequestProvider {
+	return &DynamicRequestHeaderAuthRequestProvider{
+		configmapLister: configmapLister,
+		configmapName: configmapName,
+		configmapNamespace: configmapNamespace,
+		authenticationRoleName: authenticationRoleName,
+		usernameHeadersKey: usernameHeadersKey,
+		groupHeadersKey: groupHeadersKey,
+		extraHeaderPrefixesKey: extraHeaderPrefixesKey,
+		allowedClientNamesKey: allowedClientNamesKey}
+}
+
+func (c *DynamicRequestHeaderAuthRequestProvider) UsernameHeaders() []string {
+	return c.getValuesFor(c.usernameHeadersKey)
+}
+
+func (c *DynamicRequestHeaderAuthRequestProvider) GroupHeaders() []string {
+	return c.getValuesFor(c.groupHeadersKey)
+}
+
+func (c *DynamicRequestHeaderAuthRequestProvider) ExtraHeaderPrefixes() []string {
+	return c.getValuesFor(c.extraHeaderPrefixesKey)
+}
+
+func (c *DynamicRequestHeaderAuthRequestProvider) AllowedClientNames() []string {
+	return c.getValuesFor(c.allowedClientNamesKey)
+}
+
+func (c *DynamicRequestHeaderAuthRequestProvider) getValuesFor(key string) []string {
+	configMap, err := c.configmapLister.Get(c.configmapName)
+	switch {
+	case errors.IsNotFound(err):
+		// ignore, authConfigMap is nil now
+		return nil
+	case errors.IsForbidden(err):
+		klog.Warningf("Unable to get configmap/%s in %s.  Usually fixed by "+
+			"'kubectl create rolebinding -n %s ROLEBINDING_NAME --role=%s --serviceaccount=YOUR_NS:YOUR_SA'",
+			c.configmapName, c.configmapNamespace, c.configmapNamespace, c.authenticationRoleName)
+		return nil
+	case err != nil:
+		klog.Warningf("failed to read %q configmap: %v", c.configmapName, err)
+		return nil
+	}
+
+	ret, err := deserializeStrings(configMap.Data[key])
```

Comment:
it needs to deserialize data on every request but that should be fast

Reply:
> it needs to deserialize data on every request but that should be fast

not fast enough for auth calls. We make multiple of these for every request.